src/core/lib/security/transport/secure_endpoint.cc

```diff
@@ -249,8 +252,17 @@ static void on_read(void* user_data, grpc_error_handle error) {
 
     if (ep->zero_copy_protector != nullptr) {
       // Use zero-copy grpc protector to unprotect.
-      result = tsi_zero_copy_grpc_protector_unprotect(
-          ep->zero_copy_protector, &ep->source_buffer, ep->read_buffer);
+      int last_incomplete_frame_size = 1;
+      // Get the size of the last frame which is not yet fully decrypted.
+      // This estimated is strored in ep->min_progress_size which is
+      // passed to the specify layer to indicate the minimum number of
+      // bytes that need to be read to make meaningful progress. This would
+      // avoid reading of small slices from the network.
+      result = tsi_zero_copy_grpc_protector_unprotect_get_frame_size(
+          ep->zero_copy_protector, &ep->source_buffer, ep->read_buffer,
+          &last_incomplete_frame_size);
+      last_incomplete_frame_size = std::max(1, last_incomplete_frame_size);
+      ep->min_progress_size = result != TSI_OK ? 1 : last_incomplete_frame_size;
     } else {
       // Use frame protector to unprotect.
```

Comment:
maybe add a TODO to do this for non-zero copy?

Reply:
Added a TODO